File: allocation.c
/* Memory allocation is done via functions exported from memory.c. */

/**  Allocates a pairlist.
  *
  *  @note With 'let x = f () in x::(aux y)', you have tail recursion
  *        because of the way the :: constructor is used in OCaml. You
  *        cannot have tail-recursion with R CONS and the same code.
  *        Hence the need to pre-allocate with this function.
  *  @param i Size of the pairlist to allocate.
  *  @return A newly allocated pairlist.
  */
CAMLprim value ocamlr_alloc_list (value i) {
  /* allocList is a macro wrapping up a call to the
     Rf_allocList symbol from libR.so. */
  return(Val_sexp(allocList(Int_val(i))));
}


/**  Allocates a logical vector.
  *
  *  @param i Size of the logical vector to allocate.
  *  @return A newly allocated logical vector.
  */
CAMLprim value ocamlr_alloc_lgl_vector (value i) {
  return(Val_sexp(allocVector(LGLSXP, Int_val(i))));
}


/**  Allocates a vector of integers.
  *
  *  @param i Size of the vector of integers to allocate.
  *  @return A newly allocated vector of integers.
  */
CAMLprim value ocamlr_alloc_int_vector (value i) {
  return(Val_sexp(allocVector(INTSXP, Int_val(i))));
}


/**  Allocates a vector of real numbers.
  *
  *  @param i Size of the vector of real numbers to allocate.
  *  @return A newly allocated vector of real numbers.
  */
CAMLprim value ocamlr_alloc_real_vector (value i) {
  return(Val_sexp(allocVector(REALSXP, Int_val(i))));
}


/**  Allocates a vector of strings.
  *
  *  @param i Size of the vector of strings to allocate.
  *  @return A newly allocated vector of strings.
  */
CAMLprim value ocamlr_alloc_str_vector (value i) {
  return(Val_sexp(allocVector(STRSXP, Int_val(i))));
}

File: allocation.ml
(* Memory allocation is done via functions exported from memory.c. *)


(**  Allocates a pairlist.
  *
  *  r_alloc_list takes an integer i as argument, and returns an R
  *  pairlist, also called a LISTSXP, of size i.
  *
  *  Such a function is necessary in order to avoid the tail
  *  recursion issue. While doing 'let x = f () in x::(aux y)' is
  *  an acceptable way to make aux tail recurse in Objective Caml,
  *  this relies on the way Objective Caml constructs list with
  *  the :: constructor. This feature is not available when we use
  *  stub code around the CONS macro for R. Allocating a list and
  *  filling it in imperatively is a workaround.
  *)

external alloc_list : int -> 'a lisplist sxp = "r_alloc_list"


(**  Allocates a logical vector.
  *
  *  alloc_lgl_vector takes an integer i as argument, and returns
  *  an R vector of logical values of size i.
  *)

external alloc_lgl_vector : int -> vec_lgl sxp = "r_alloc_lgl_vector"


(**  Allocates a vector of integers.
  *
  *  alloc_int_vector takes an integer i as argument, and returns
  *  an R vector of integer values of size i.
  *)

external alloc_int_vector : int -> vec_int sxp = "r_alloc_int_vector"


(**  Allocates a vector of strings.
  *
  *  r_alloc_str_vector takes an integer i as argument, and returns
  *  an R vector of i strings.
  *)

external alloc_str_vector : int -> vec_str sxp = "r_alloc_str_vector"

